worker.go

```diff
@@ -330,47 +330,45 @@ func (ew *EventWorker) Flush() ([]samplers.UDPEvent, []samplers.UDPServiceCheck)
 	return retevts, retsvchecks
 }
 
-// TraceWorker is similar to a Worker but it collects events and service checks instead of metrics.
-type TraceWorker struct {
-	TraceChan  chan ssf.SSFSpan
-	mutex      *sync.Mutex
-	traces     *ring.Ring
-	stats      *statsd.Client
-	bufferSize int
+// SpanWorker is similar to a Worker but it collects events and service checks instead of metrics.
+type SpanWorker struct {
+	SpanChan chan ssf.SSFSpan
+	sinks    []spanSink
+	stats    *statsd.Client
 }
 
-// NewTraceWorker creates an TraceWorker ready to collect events and service checks.
-func NewTraceWorker(stats *statsd.Client, bufferSize int) *TraceWorker {
-	return &TraceWorker{
-		TraceChan:  make(chan ssf.SSFSpan),
-		mutex:      &sync.Mutex{},
-		traces:     ring.New(bufferSize),
-		stats:      stats,
-		bufferSize: bufferSize,
+// NewSpanWorker creates an SpanWorker ready to collect events and service checks.
+func NewSpanWorker(sinks []spanSink, stats *statsd.Client) *SpanWorker {
+	return &SpanWorker{
+		SpanChan: make(chan ssf.SSFSpan),
+		sinks:    sinks,
+		stats:    stats,
 	}
 }
 
-// Work will start the EventWorker listening for events and service checks.
+// Work will start the SpanWorker listening for spans.
 // This function will never return.
-func (tw *TraceWorker) Work() {
-	for m := range tw.TraceChan {
-		tw.mutex.Lock()
-		tw.traces.Value = m
-		tw.traces = tw.traces.Next()
-		tw.mutex.Unlock()
+func (tw *SpanWorker) Work() {
+	for m := range tw.SpanChan {
+		// Give each sink a change to ingest.
+		for _, s := range tw.sinks {
+			err := s.Ingest(m)
+			if err != nil {
+				// If a sink goes wacko and errors a lot, we stand to emit a loooot of metrics
+				// here since span ingest rates can be very high. C'est la vie.
+				tw.stats.Count("worker.span.ingest_error_total", 1, []string{fmt.Sprintf("sink:%s", s.Name())}, 1.0)
+			}
+		}
 	}
 }
 
-// Flush returns the TraceWorker's stored spans and
-// resets the stored contents.
-func (tw *TraceWorker) Flush() *ring.Ring {
-	start := time.Now()
-	tw.mutex.Lock()
+// Flush invokes flush on each sink.
+func (tw *SpanWorker) Flush() {
 
-	rettraces := tw.traces
-	tw.traces = ring.New(tw.bufferSize)
-
-	tw.mutex.Unlock()
-	tw.stats.TimeInMilliseconds("flush.event_worker_duration_ns", float64(time.Since(start).Nanoseconds()), nil, 1.0)
-	return rettraces
+	// Flush and time each sink.
+	for _, s := range tw.sinks {
+		sinkFlushStart := time.Now()
+		s.Flush()
+		tw.stats.TimeInMilliseconds("worker.span.flush_duration_ns", float64(time.Since(sinkFlushStart).Nanoseconds()), []string{fmt.Sprintf("sink:%s", s.Name())}, 1.0)
```

Comment:
Just a callout that this may require updating the dashboard as part of the rollout